agent/immortalstreams/stream.go

```diff
@@ -0,0 +1,253 @@
+package immortalstreams
+
+import (
+	"context"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/v2/agent/agentssh"
+	"github.com/coder/coder/v2/agent/immortalstreams/backedpipe"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+// Stream represents an immortal stream connection
+type Stream struct {
+	id        uuid.UUID
+	name      string
+	port      uint16
+	createdAt time.Time
+	logger    slog.Logger
+
+	mu                  sync.RWMutex
+	localConn           io.ReadWriteCloser
+	pipe                *backedpipe.BackedPipe
+	lastConnectionAt    time.Time
+	lastDisconnectionAt time.Time
+	closed              bool
+
+	// goroutines manages the copy goroutines
+	goroutines sync.WaitGroup
+
+	// Shutdown signal
+	shutdownChan chan struct{}
+
+	// Context cancellation for BackedPipe
+	cancel context.CancelFunc
+}
+
+// NewStream creates a new immortal stream
+func NewStream(id uuid.UUID, name string, port uint16, logger slog.Logger) *Stream {
+	// Create a context that will be canceled when the stream is closed
+	ctx, cancel := context.WithCancel(context.Background())
+
+	stream := &Stream{
+		id:           id,
+		name:         name,
+		port:         port,
+		createdAt:    time.Now(),
+		logger:       logger,
+		shutdownChan: make(chan struct{}),
+		cancel:       cancel, // Store cancel function for cleanup
+		// Create BackedPipe without a reconnector; reconnections are accepted
+		// explicitly via HandleReconnect.
+		pipe: backedpipe.NewBackedPipe(ctx, nil),
+	}
+
+	// Track disconnection time via BackedPipe callback rather than read/write errors.
+	stream.pipe.SetDisconnectedCallback(func() {
+		stream.handleDisconnect()
+	})
+
+	return stream
+}
+
+// setNameAndLogger sets the stream name and updates the logger to include it.
+// Must be called by the manager under its own lock before publishing the stream.
+func (s *Stream) setNameAndLogger(name string, baseLogger slog.Logger) {
+	s.mu.Lock()
+	s.name = name
+	s.logger = baseLogger.With(slog.F("stream_name", name))
+	s.mu.Unlock()
+}
+
+// Start starts the stream with an initial connection
+func (s *Stream) Start(localConn io.ReadWriteCloser) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.closed {
+		return xerrors.New("stream is closed")
+	}
+
+	s.localConn = localConn
+	s.lastConnectionAt = time.Now()
```

Comment:
the local connection is not the connection that `lastConnectionAt` is about. The local connection has the same lifetime as the stream itself. `lastConnectionAt` and `lastDisconnectionAt` are about the _unreliable_ connections, which come and go.